crates/core/src/error/summary.rs

```diff
@@ -0,0 +1,617 @@
+    pub fn complete(&mut self) {
+        self.end_time = Some(Instant::now());
+    }
+
+    pub fn enable_log(&mut self) -> Self {
+        self.log_enabled = true;
+        self.clone()
+    }
+
+    pub fn contains_error(&self) -> bool {
+        self.issues.contains_key(&IssueCategory::Error)
+            && !self.issues[&IssueCategory::Error].is_empty()
+    }
+
+    pub fn add_error(
+        &mut self,
+        scope: IssueScope,
+        message: impl Into<EcoString>,
+        root_cause: impl Into<Option<EcoString>>,
+    ) {
+        self.add_issue(IssueCategory::Error, scope, message, root_cause);
+    }
+
+    pub fn add_warning(
+        &mut self,
+        scope: IssueScope,
+        message: impl Into<EcoString>,
+        root_cause: impl Into<Option<EcoString>>,
+    ) {
+        self.add_issue(IssueCategory::Warning, scope, message, root_cause);
+    }
+
+    pub fn add_info(
+        &mut self,
+        scope: IssueScope,
+        message: impl Into<EcoString>,
+        root_cause: impl Into<Option<EcoString>>,
+    ) {
+        self.add_issue(IssueCategory::Info, scope, message, root_cause);
+    }
+
+    /// Adds a new issue to the summary, condensing similar issues
+    pub fn add_issue(
+        &mut self,
+        category: IssueCategory,
+        scope: IssueScope,
+        message: impl Into<EcoString>,
+        root_cause: impl Into<Option<EcoString>>,
+    ) {
+        let root_cause = root_cause.into();
+        let message = message.into();
+
+        if self.log_enabled {
+            // We ignore the result here, as we don't want to propagate the error
+            _ = Self::log_issue(scope, category, &message, &root_cause);
+        }
+
+        _ = self
+            .issues
+            .entry(category)
+            .or_default()
+            .entry(scope)
+            .and_modify(|val| {
+                val.count += 1;
+                if val.root_cause.is_none() {
+                    val.root_cause.clone_from(&root_cause);
+                }
+            })
+            .or_insert(CondensedIssue {
+                category,
+                message,
+                count: 1,
+                root_cause,
+            });
+    }
+
+    /// Adds a custom metric
+    pub fn add_metric(&mut self, key: impl Into<EcoString>, value: impl Into<EcoString>) {
+        let value = value.into();
+
+        _ = self
+            .metrics
+            .entry(key.into())
+            .and_modify(|val| *val = value.clone())
+            .or_insert_with(|| value);
+    }
+
+    pub fn export_issues(&mut self) -> bool {
+        self.display.insert(DisplayOptionKind::Issues)
+    }
+
+    pub fn export_timing(&mut self) -> bool {
+        self.display.insert(DisplayOptionKind::Timing)
+    }
+
+    pub fn export_metrics(&mut self) -> bool {
+        self.display.insert(DisplayOptionKind::Metrics)
+    }
+
+    pub fn export_all(&mut self) -> bool {
+        self.display.insert(DisplayOptionKind::All)
+    }
```

Comment:
could you let me know of use cases where we want only some `DisplayOptionKind` and not some others?


Also I feel it's a bit weird that `ALL`  is an enum value.